net/multi_listen.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+)
+
+// connErrIndexTriplet packs results of accept call on actual listeners along with the
+// index of the listener. It is used for communication between main and listener goroutines.
+type connErrIndexTriplet struct {
+	index int
+	conn  net.Conn
+	err   error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	latestAcceptedIndex atomic.Int32
+	addrs               []net.Addr
+	listeners           []net.Listener
+
+	wg             sync.WaitGroup
+	connErrIndexCh chan connErrIndexTriplet
+	closed         atomic.Bool
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(ctx context.Context, addresses []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		addresses,
+		net.ResolveTCPAddr,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
```

Comment:
I assume the "Func"s are there to simplify unit-test? If so, please comment that since it otherwise seems ... un-motivated

Reply:
Also, please verify that when net.ListenConfig is used, then the "KeepAlive" default is the golang-default (15s), and not the os-default (usually 2h on Linux). If it is the os-default, then set it to 15s explicitly to be in sync with golang standard, but that may of course change in the future (sigh),

Reply:
https://github.com/golang/go/blob/go1.22.4/src/net/dial.go#L806-L809
net.Listen() starts with empty ListenConfig, all the default values will besame.